pkgs/applications/virtualization/xen/4.18/4.18.nix

```diff
@@ -0,0 +1,194 @@
+{
+  lib,
+  callPackage,
+  fetchpatch,
+  fetchgit,
+  fetchFromGitHub,
+  ocaml-ng,
+
+  withInternalQEMU ? true,
+
+  withInternalTraditionalQEMU ? true,
+
+  withInternalSeaBIOS ? true,
+  withSeaBIOS ? !withInternalSeaBIOS,
+  seabios,
+
+  withInternalOVMF ? true,
+  withOVMF ? !withInternalOVMF,
+  OVMF,
+
+  withInternalIPXE ? true,
+  withIPXE ? !withInternalIPXE,
+  ipxe,
+
+  # QEMU
+  udev,
+  pciutils,
+  xorg,
+  SDL,
+  pixman,
+  acl,
+  glusterfs,
+  spice-protocol,
+  usbredir,
+  alsa-lib,
+  glib,
+  python3,
+
+  # OVMF
+  nasm,
+  bash,
+  ...
+}@args:
+
+assert withInternalSeaBIOS -> !withSeaBIOS;
+assert withInternalOVMF -> !withOVMF;
+assert withInternalIPXE -> !withIPXE;
+
+with lib;
+
+let
+  upstreamPatches = import ../patches.nix { inherit fetchpatch; };
+
+  qemuDeps = [
+    udev
+    pciutils
+    xorg.libX11
+    SDL
+    pixman
+    acl
+    glusterfs
+    spice-protocol
+    usbredir
+    alsa-lib
+    glib
+    python3
+  ];
+in
+
+callPackage (import ../generic.nix rec {
+  version = "4.18.2";
+  qemuVersion = "4.18.1";
+  seaBIOSVersion = "1.16.3";
+  ovmfVersion = "202305";
+
+  src = fetchgit {
+    url = "git://xenbits.xenproject.org/xen.git";
+    rev = "RELEASE-${version}";
+    hash = "sha256-pHCjj+Bcy4xQfB9xHU9fccFwVdP2DXrUhdszwGvrdmY=";
+  };
+
+  # Sources needed to build tools and firmwares.
+  prefetchedSources =
+    optionalAttrs withInternalQEMU {
+      qemu-xen = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen.git";
+          rev = "qemu-xen-${qemuVersion}";
+          hash = "sha256-BX+LXfNzwdUMALwwI1ZDW12dJ357oynjnrboLHREDGQ=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace scripts/tracetool.py \
+            --replace-fail "/usr/bin/env python" "${python3}/bin/python"
+        '';
+        meta.description = "Xen's fork of upstream QEMU";
```

Comment:
Would this conflict with `boot.binfmt.emulatedSystems`?

Reply:
Currently Xen (and its related components) only build on x86_64-linux, as we use a custom build of binutils to enable EFI building on i386pep. This has been the case since radhus' update to 4.15 (though the Makefile patches were broken even then), but Xen on nixpkgs has been x86-only since it was introduced.
I hope to eventually expand the build to aarch64 to also allow for Xen dom0less and (eventually) Hyperlaunch support, but for now, I wouldn't emulate or build this to any other platform.

Reply:
I meant more for https://github.com/NiXium-org/NiXium where binfmt is used for distributed builds and to send instructions to the infra to schedule a build for non-natives -> So if xen-enabled system can still perform distributed builds as it's using a qemu fork.